internal/go118_sha1_patch.go

```diff
@@ -0,0 +1,47 @@
+package internal
+
+import (
+	"fmt"
+	"os"
+	"sync"
+	_ "unsafe" // for go:linkname
+
+	goversion "github.com/hashicorp/go-version"
+	"github.com/hashicorp/vault/sdk/version"
+)
+
+const sha1PatchVersionsBefore = "1.12.0"
+
+var patchSha1 sync.Once
+
+//go:linkname debugAllowSHA1 crypto/x509.debugAllowSHA1
+var debugAllowSHA1 bool
+
+// PatchSha1 patches Go 1.18+ to allow certificates with signatures containing SHA-1 hashes to be allowed.
+// It is safe to call this function multiple times.
+// This is necessary to allow Vault 1.10 and 1.11 to work with Go 1.18+ without breaking backwards compatibility
+// with these certificates. See https://go.dev/doc/go1.18#sha1 and
+// https://developer.hashicorp.com/vault/docs/deprecation/faq#q-what-is-the-impact-of-removing-support-for-x-509-certificates-with-signatures-that-use-sha-1
+// for more details.
+// TODO: remove when Vault <=1.11 is no longer supported
+func PatchSha1() {
+	patchSha1.Do(func() {
+		patchBefore, err := goversion.NewSemver(sha1PatchVersionsBefore)
+		if err != nil {
+			panic(err)
```

Comment:
Are we sure we want a panic here? Later you set `patch = true` if `err` is not `nil`.

Reply:
`sha1PatchVersionsBefore` is a `const`, so this should always succeed. If it doesn't, I think we'd rather know about it during unit tests than have it fail.

Reply:
Fair. But then why not panic in the `else` on line 38?

Reply:
Mostly because that value isn't a `const`, since it isn't set until we build and link the binary. I could panic there too. I just slightly prefer to fail a little more safely if the user sets the version incorrectly.